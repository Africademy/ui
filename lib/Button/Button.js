"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.ButtonStyled = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var Styles = _interopRequireWildcard(require("./style"));

var _ChevronDown = _interopRequireDefault(require("../Icon/Icons/ChevronDown"));

var _Select = _interopRequireDefault(require("../Select/Select"));

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _templateObject3() {
  var data = _taggedTemplateLiteral(["\n  position: absolute;\n  overflow: hidden;\n  clip: rect(0 0 0 0);\n  height: 1px; width: 1px;\n  margin: -1px; padding: 0; border: 0;\n"]);

  _templateObject3 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = _taggedTemplateLiteral(["\n  width: 24px;\n  margin-left: 10px;\n"]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = _taggedTemplateLiteral(["\n  ", ";\n  ", ";\n  ", ";\n  ", "\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

var ButtonStyled = _styledComponents.default.button(_templateObject(), Styles.ButtonBase, function (props) {
  return Styles[props.size];
}, function (props) {
  return Styles[props.disabled ? 'disabled' : props.type];
}, function (props) {
  return Styles[props.fullWidth ? 'fullWidth' : ''];
});

exports.ButtonStyled = ButtonStyled;

var Loading = _styledComponents.default.img(_templateObject2());

var VisuallyHiddenLabel = _styledComponents.default.span(_templateObject3());
/** All buttons, including text, link and split-buttons, follow the same core principles in dimensions, padding, and font sizes.
 * Combined with simple modifiers, they can be changed in size and appearance.  */


var Button = function Button(_ref) {
  var disabled = _ref.disabled,
      onClick = _ref.onClick,
      type = _ref.type,
      size = _ref.size,
      label = _ref.label,
      isSplit = _ref.isSplit,
      loading = _ref.loading,
      icon = _ref.icon,
      hasIconOnly = _ref.hasIconOnly,
      isSelect = _ref.isSelect,
      items = _ref.items,
      selectPosition = _ref.selectPosition,
      onSelectClick = _ref.onSelectClick,
      fullWidth = _ref.fullWidth;
  return _react.default.createElement(ButtonStyled, {
    onClick: !disabled ? onClick : undefined,
    disabled: disabled,
    size: size,
    type: type,
    isSplit: isSplit,
    icon: icon,
    hasIconOnly: hasIconOnly,
    fullWidth: fullWidth
  }, icon, hasIconOnly && _react.default.createElement(VisuallyHiddenLabel, null, label), !hasIconOnly && _react.default.createElement(Styles.ButtonLabel, {
    hasIcon: !!icon
  }, label), isSelect && (type === 'primary' || type === 'secondary') && _react.default.createElement(Styles.ButtonArrow, null, _react.default.createElement(_ChevronDown.default, {
    color: type === 'primary' ? 'white' : 'grayDark',
    size: size,
    isChevron: true
  })), loading && _react.default.createElement(Loading, {
    src: "./images/loading-gray.svg",
    alt: "loading"
  }), isSplit && (type === 'primary' || type === 'secondary') && _react.default.createElement(_Select.default, {
    onSelectClick: onSelectClick,
    items: items,
    type: type,
    isSplit: true,
    position: selectPosition,
    disabled: disabled
  }));
};

Button.propTypes = {
  /** Is the button disabled */
  disabled: _propTypes.default.bool,

  /** Size of the Button */
  size: _propTypes.default.oneOf(['small', 'large', 'medium']),

  /** OnClick handler */
  onClick: _propTypes.default.func,

  /** Button label */
  label: _propTypes.default.string,

  /** Type of button */
  type: _propTypes.default.oneOf(['link', 'primary', 'secondary', 'text']),

  /** Is the Button Split  */
  isSplit: _propTypes.default.bool,

  /** Is this the Select button with chevron */
  isSelect: _propTypes.default.bool,

  /** Is the Button Loading  */
  loading: _propTypes.default.bool,

  /** Does the button have only an icon and no label */
  hasIconOnly: _propTypes.default.bool,

  /** Icon to show with the label */
  icon: _propTypes.default.node,

  /** Items to display in the Split Button popup */
  items: _propTypes.default.arrayOf(_propTypes.default.shape({
    id: _propTypes.default.string.isRequired,
    title: _propTypes.default.string.isRequired
  })),

  /** Position of the Select popup */
  selectPosition: _propTypes.default.oneOf(['top', 'bottom']),

  /** Function to call on Split Button selected item click */
  onSelectClick: _propTypes.default.func,

  /** Is the button the full width of the parent container */
  fullWidth: _propTypes.default.bool
};
Button.defaultProps = {
  disabled: false,
  isSplit: undefined,
  loading: false,
  size: 'medium',
  type: 'secondary',
  label: undefined,
  hasIconOnly: false,
  icon: undefined,
  isSelect: undefined,
  items: undefined,
  selectPosition: 'bottom',
  onSelectClick: undefined,
  fullWidth: false,
  onClick: null
};
var _default = Button;
exports.default = _default;