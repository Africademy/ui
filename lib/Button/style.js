"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ButtonLabel = exports.ButtonArrow = exports.ButtonSelect = exports.fullWidth = exports.disabled = exports.large = exports.medium = exports.small = exports.text = exports.link = exports.secondary = exports.primary = exports.ButtonBase = void 0;

var _styledComponents = _interopRequireWildcard(require("styled-components"));

var _colors = require("../style/colors");

var _fonts = require("../style/fonts");

var _borders = require("../style/borders");

var _Button = require("./Button");

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _templateObject13() {
  var data = _taggedTemplateLiteral(["\n  margin-left: ", ";\n"]);

  _templateObject13 = function _templateObject13() {
    return data;
  };

  return data;
}

function _templateObject12() {
  var data = _taggedTemplateLiteral(["\n  padding-left: 5px;\n  display: flex;\n"]);

  _templateObject12 = function _templateObject12() {
    return data;
  };

  return data;
}

function _templateObject11() {
  var data = _taggedTemplateLiteral(["\n  :before {\n    background-color: ", ";\n    color: ", ";\n    content: \"\";\n    height: 24px;\n    position: absolute;\n    right: 34px;\n    top: 50%;\n    transform: translateY(-50%);\n    width: 1px;\n  }\n  padding-left: 13px;\n  display: flex;\n  justify-content: center;\n  align-items: center;\n  width: 100%;\n  height: 38px;\n  ", ":hover & {\n    background-color: ", ";\n    color: ", ";\n  }\n  ", ":focus & {\n    border-color: ", ";\n      color: ", ";\n  }\n"]);

  _templateObject11 = function _templateObject11() {
    return data;
  };

  return data;
}

function _templateObject10() {
  var data = _taggedTemplateLiteral(["\n  width: 100%;\n  justify-content: center;\n"]);

  _templateObject10 = function _templateObject10() {
    return data;
  };

  return data;
}

function _templateObject9() {
  var data = _taggedTemplateLiteral(["\n  background-color: ", ";\n  cursor: not-allowed;\n"]);

  _templateObject9 = function _templateObject9() {
    return data;
  };

  return data;
}

function _templateObject8() {
  var data = _taggedTemplateLiteral(["\n  padding-right: 16px;\n  padding-left: 16px;\n  height: 48px;\n"]);

  _templateObject8 = function _templateObject8() {
    return data;
  };

  return data;
}

function _templateObject7() {
  var data = _taggedTemplateLiteral(["\n  padding-right: 16px;\n  padding-left: 16px;\n  height: 40px;\n"]);

  _templateObject7 = function _templateObject7() {
    return data;
  };

  return data;
}

function _templateObject6() {
  var data = _taggedTemplateLiteral(["\n  padding-left: 8px;\n  padding-right: 8px;\n  height: 32px;\n"]);

  _templateObject6 = function _templateObject6() {
    return data;
  };

  return data;
}

function _templateObject5() {
  var data = _taggedTemplateLiteral(["\n  background-color: transparent;\n  :hover {\n    color: ", ";\n  }\n\n  :focus,\n  :active {\n    box-shadow: none;\n    outline: 0;\n  }\n"]);

  _templateObject5 = function _templateObject5() {
    return data;
  };

  return data;
}

function _templateObject4() {
  var data = _taggedTemplateLiteral(["\n  background-color: transparent;\n  color: ", ";\n  :hover {\n    color: ", ";\n  }\n"]);

  _templateObject4 = function _templateObject4() {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = _taggedTemplateLiteral(["\n  background-color: ", ";\n  color: ", ";\n  border: 1px solid ", ";\n\n  :hover {\n    color: ", ";\n    border-color: ", ";\n  }\n\n  :focus {\n    border-color: ", ";\n    color: ", ";\n  }\n"]);

  _templateObject3 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = _taggedTemplateLiteral(["\n  background-color: ", ";\n  color: ", ";\n  :hover {\n    background-color: ", ";\n    color: ", ";\n  }\n"]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = _taggedTemplateLiteral(["\n  display: flex;\n  align-items: center;\n  justify-content: flex-start;\n  font-size: ", ";\n  line-height: ", ";\n  font-weight: ", ";\n  border-radius: ", ";\n  cursor: pointer;\n  user-select: none;\n  border: 0;\n  -webkit-appearance: none;\n  position: relative;\n  transition-property: background-color, border-color, color;\n  transition-duration: 0.1s;\n  transition-timing-function: ease-in;\n  white-space: nowrap;\n  text-overflow: ellipsis;\n  min-width: 0;\n  flex: 0 0 auto;\n  font-family: ", ";\n  background-color: ", ";\n  color: ", ";\n  padding-top: 0;\n  padding-bottom: 0;\n  :hover {\n    transition-property: background-color, border-color, color;\n    transition-duration: 0.1s;\n    transition-timing-function: ease-in;\n  }\n  :active {\n    top: 1px;\n  }\n  :focus {\n    box-shadow: 0 0 0 3px ", ";\n    outline: 0;\n  }\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

/* styles common to all buttons */
var ButtonBase = (0, _styledComponents.css)(_templateObject(), _fonts.fontSize, _fonts.lineHeight, _fonts.fontWeightMedium, _borders.borderRadius, _fonts.fontFamily, _colors.white, _colors.grayDefault, _colors.boxShadow);
/* theme variants */

exports.ButtonBase = ButtonBase;
var primary = (0, _styledComponents.css)(_templateObject2(), _colors.blue, _colors.white, _colors.blueDarker, _colors.white);
exports.primary = primary;
var secondary = (0, _styledComponents.css)(_templateObject3(), _colors.white, _colors.grayDark, _colors.gray, _colors.grayDarker, _colors.grayDark, _colors.blue, _colors.grayDarker);
exports.secondary = secondary;
var link = (0, _styledComponents.css)(_templateObject4(), _colors.blue, _colors.blueDarker);
exports.link = link;
var text = (0, _styledComponents.css)(_templateObject5(), _colors.grayDarker);
/* size variants */

exports.text = text;
var small = (0, _styledComponents.css)(_templateObject6());
exports.small = small;
var medium = (0, _styledComponents.css)(_templateObject7());
exports.medium = medium;
var large = (0, _styledComponents.css)(_templateObject8());
/* state variants */

exports.large = large;
var disabled = (0, _styledComponents.css)(_templateObject9(), _colors.grayLight);
exports.disabled = disabled;
var fullWidth = (0, _styledComponents.css)(_templateObject10());
exports.fullWidth = fullWidth;

var ButtonSelect = _styledComponents.default.div(_templateObject11(), function (props) {
  return props.disabled ? _colors.gray : _colors.white;
}, function (props) {
  return props.disabled ? _colors.gray : _colors.white;
}, _Button.ButtonStyled, _colors.blueDarker, _colors.white, _Button.ButtonStyled, _colors.blue, _colors.grayDarker);

exports.ButtonSelect = ButtonSelect;

var ButtonArrow = _styledComponents.default.div(_templateObject12());

exports.ButtonArrow = ButtonArrow;

var ButtonLabel = _styledComponents.default.div(_templateObject13(), function (props) {
  return props.hasIcon ? '5px' : '0px';
});

exports.ButtonLabel = ButtonLabel;