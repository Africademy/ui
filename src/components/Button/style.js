import styled from 'styled-components';

export const ButtonStyled = styled.div`
  height:  ${props =>
    (props.large ? '48px' :
        props.small ? '32px' :
                '40px')};
  display: flex;
  align-items: center;
  justify-content: flex-start;
  padding-left: ${props=> (props.small ? '8px' : '16px')};
  padding-right:  ${props=> (props.small ? '8px' : '16px')};
  font-size: 14px;
  line-height: 14px;
  font-weight: 500;
  border-radius: 4px;
  cursor: pointer;
  user-select: none;
  border: 0;
  -webkit-appearance: none;
  position: relative;
  transition-property: background-color, border-color, color;
  transition-duration: 0.1s;
  transition-timing-function: ease-in;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  min-width: 0;
  flex: 0 0 auto;
  font-family: Roboto,serif;
  background-color: ${props =>
    (props.primary ? '#2c4bff' :
        props.secondary ? '#fff' :
            props.link || props.text ? 'transparent' :
                'white')};
  color: ${props =>
    (props.primary ? '#fff' :
        props.secondary ? '636363' :
            props.link ? '#2C4BFF' :
                '#77797a')};
  border: ${props => (props.secondary ? '1px solid #B8B8B8' : '0')};
  
  :hover {
      transition-property: background-color, border-color, color;
      transition-duration: 0.1s;
      transition-timing-function: ease-in;
      background-color: ${props => (props.primary ? '#1f35b3' : 'inherit')};
      border-color: ${props => (props.secondary ? '#636363' : 'inherit')};
      color: ${props =>
    (props.secondary ? '#3d3d3d' :
        props.text ? '#3d3d3d' :
            props.link ? '#1F35B3' :
                '#fff')};
  }
  
  :active {
    top: 1px;
  }
  
  :focus {
      box-shadow: 0 0 0 3px #ABB7FF;
      outline: 0;
      border-color: ${props => (props.secondary ? '#2c4bff' : 'inherit')};
      color: ${props => (props.secondary ? '#3d3d3d' : 'inherit')};
    }
`;

