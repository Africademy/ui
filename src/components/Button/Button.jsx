import React from 'react';
import PropTypes from 'prop-types';
import style from 'styled-components';
import * as Styles from './style';
import ChevronDown from '../Icon/Icons/ChevronDown';
import Select from '../Select/Select';

export const ButtonStyled = style.div`  
  ${Styles.buttonbase};
  ${props => Styles[props.size]};
  ${props => Styles[props.disabled ? 'disabled' : props.type]};
`;

const Loading = style.img`
  width: 24px;
  padding-left: 10px;
`;


/** All buttons, including text, link and split-buttons, follow the same core principles in dimensions, padding, and font sizes.
 * Combined with simple modifiers, they can be changed in size and appearance.  */
const Button = ({
  disabled,
  onClick,
  type,
  size,
  label,
  isSplit,
  loading,
  icon,
  hasIconOnly,
  isSelect,
  items,
}) => (
  <ButtonStyled
    onClick={!disabled && onClick}
    disabled={disabled}
    size={size}
    type={type}
    isSplit={isSplit}
    icon={icon}
    hasIconOnly={hasIconOnly}
    aria-label={label || null}
  >
    {label}
    {isSelect && <Styles.ArrowButton><ChevronDown type={type} size={size} /></Styles.ArrowButton>}
    {loading && <Loading src="./images/loading-gray.svg" alt="loading" />}
    {isSplit && <Select onSelectClick={() => {}} items={items} type={type} isSplit /> }
  </ButtonStyled>
);


Button.propTypes = {
  /** Is the button disabled */
  disabled: PropTypes.bool,

  /** Size of the Button */
  size: PropTypes.oneOf(['small', 'large', 'medium']),

  /** OnClick handler */
  onClick: PropTypes.func.isRequired,

  /** Button label */
  label(props, propName) {
    if ((props.hasIconOnly === false && (props[propName] === undefined || typeof (props[propName]) !== 'string'))) {
      return new Error(
        'Please provide a label.',
      );
    }
  },

  /** Type of button */
  type: PropTypes.oneOf(['link', 'primary', 'secondary', 'text']),

  /** Is the Button Split  */
  isSplit: PropTypes.bool,

  /** Is this the Select button with chevron */
  isSelect: PropTypes.bool,

  /** Is the Button Loading  */
  loading: PropTypes.bool,

  /** Does the button have only an icon and no label */
  hasIconOnly: PropTypes.bool,

  /** Icon to show with the label */
  icon(props, propName) {
    if ((props.hasIconOnly === true && (props[propName] === undefined || typeof (props[propName]) !== 'string'))) {
      return new Error(
        'Please provide a label.',
      );
    }
  },

  items: PropTypes.shape({
    id: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
  }).isRequired,
};


Button.defaultProps = {
  disabled: false,
  isSplit: false,
  loading: false,
  size: 'medium',
  type: 'secondary',
  label: undefined,
  hasIconOnly: false,
  icon: undefined,
  isSelect: false,
};


export default Button;
