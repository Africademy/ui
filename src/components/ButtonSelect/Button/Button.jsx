import React from 'react';
import PropTypes from 'prop-types';
import style from 'styled-components';
import * as ButtonStyles from '../../Button/style';
import * as ButtonSplitStyles from './style';

const ButtonStyled = style.div`  
  ${ButtonStyles.buttonbase};
  ${ButtonSplitStyles.buttonSplitBase};
  ${props => ButtonStyles[props.size]};
  ${props => ButtonStyles[props.disabled ? 'disabled' : props.type]};
`;

/** Atom component used by the ButtonSplit component to render the right side Select button.
 * On click, this component opens the Select menu popup. */
const Button = ({
  onClick, type, disabled, size,
}) => (
  <ButtonStyled disabled={disabled} type={type} size={size} onClick={onClick}>
    <ButtonSplitStyles.selectButton type={type}>â–¾</ButtonSplitStyles.selectButton>
  </ButtonStyled>
);

Button.propTypes = {
  /** Function to call on Select Button click to open the menu popup */
  onClick: PropTypes.func.isRequired,

  /** Type of the Select button */
  type: PropTypes.oneOf(['primary', 'secondary']),

  /** Is the button disabled */
  disabled: PropTypes.bool,

  /** Size of the Select button */
  size: PropTypes.string,
};

Button.defaultProps = {
  disabled: false,
  size: 'medium',
  type: 'secondary',
};

export default Button;
