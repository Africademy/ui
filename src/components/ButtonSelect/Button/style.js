import style, { css } from 'styled-components';
import {
  blueLighter, gray,
} from '../../style/colors';

export const buttonSplitBase = css`
  overflow: visible;
  border-top-left-radius: 0;
  border-bottom-left-radius: 0;
`;

export const selectButton = style.div`
  :before {
    background-color: ${props => (props.primary ? blueLighter : gray)};
    content: "";
    height: 24px;
    position: absolute;
    right: 38px;
    top: 50%;
    transform: translateY(-50%);
    width: 1px;
  }
`;
