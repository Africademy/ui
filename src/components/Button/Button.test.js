import snap from 'jest-auto-snapshots';
import Button from './Button';

snap(Button, './Button.jsx');