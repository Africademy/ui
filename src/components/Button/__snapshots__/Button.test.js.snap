// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly 1`] = `
<div
  aria-label={null}
  className="sc-bdVaJa fThgCj"
  disabled={false}
  onClick={[Function]}
>
  Primary Button
</div>
`;
